src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/collator/Collator.java

```diff
@@ -0,0 +1,148 @@
+ * order for the above guarantee to work, there should be only one collator instance in an {@link
+ * AnalysisGraph}.
+ */
+public class Collator extends Decider {
+
+  public static final String NAME = "collator";
+
+  /* Deciders can choose to publish decisions at different frequencies based on the
+   * type of resources monitored and rca signals. The collator should however, not introduce any
+   * unnecessary delays. As soon as a decision is taken, it should be evaluated and published downstream.
+   */
+  private static final int collatorFrequency = 1; // Measured in terms of number of evaluationIntervalPeriods
+
+  private static final int evalIntervalSeconds = 5;
+
+  private final List<Decider> deciders;
+
+  private final ActionGrouper actionGrouper;
+
+  public Collator(Decider... deciders) {
+    this(new SingleNodeImpactActionGrouper(), deciders);
+  }
+
+  public Collator(ActionGrouper actionGrouper, Decider... deciders) {
+    super(evalIntervalSeconds, collatorFrequency);
+    this.deciders = Arrays.asList(deciders);
+    this.actionGrouper = actionGrouper;
+  }
+
+  @Override
+  public String name() {
+    return NAME;
+  }
+
+  /**
+   * The collator uses an action grouping strategy to first group actions by instanceIds. Then, the
+   * collator polarizes the list of actions per instance to be in the same direction of pressure,
+   * i.e. all the polarized actions either increase pressure on a node, or decrease pressure on a
+   * node.
+   *
+   * <p>When there are conflicting actions suggested by the deciders for an instance, the
+   * polarization logic prefers pruning actions that decrease stability retaining only those that
+   * increase stability. </p>
+   *
+   * @return A {@link Decision} instance that contains the list of polarized actions.
+   */
+  @Override
+  public Decision operate() {
+    final List<Action> proposedActions = getAllProposedActions();
+    final Map<NodeKey, List<Action>> actionsByNode = actionGrouper
+        .groupByInstanceId(proposedActions);
+    final List<Action> prunedActions = new ArrayList<>();
+    actionsByNode.forEach((nodeKey, actions) -> prunedActions.addAll(polarize(nodeKey, actions)));
+
+    final Decision finalDecision = new Decision(System.currentTimeMillis(), NAME);
+    finalDecision.addAllActions(prunedActions);
+    return finalDecision;
+  }
+
+  @NonNull
+  private List<Action> getAllProposedActions() {
+    final List<Action> proposedActions = new ArrayList<>();
+    if (deciders != null) {
+      for (final Decider decider : deciders) {
+        List<Decision> decisions = decider.getFlowUnits();
+        if (decisions != null) {
+          decisions.forEach(decision -> {
+            if (decision.getActions() != null) {
+              proposedActions.addAll(decision.getActions());
+            }
+          });
+        }
+      }
+    }
+    return proposedActions;
+  }
+
+  private List<Action> polarize(final NodeKey nodeKey, final List<Action> actions) {
```

Comment:
It seems to me that this will need non-trivial changes when we have actions impacting multiple nodes. If an action is purged because of pressure conflict on one of the nodes, it must be removed from the list of all of the nodes. 

Should `polarize()` then move to `ActionGrouper`? Is the overall above structure right for future use cases?

---

My ideas for solving this were more around compiling all impact vectors into a matrix and doing a second pass through it to accept or prune actions. Something on these lines:

#### Suppose the collator received 2 actions, moveShard from node A to node B, and moveShard from node B to node C
```ruby
actions = [MoveShard: A -> B, MoveShard: B -> C]
Impact vectors:
	actions[0] :: MoveShard: A -> B
	impactedNodes = [A, B]
	A: [CPU:dec, Heap: dec, RAM: dec, NW: dec, Disk: dec]
	B: [CPU:inc, Heap: inc, RAM: inc, NW: inc, Disk: inc]

	actions[1] :: MoveShard: B -> C
	impactedNodes = [B, C]
	B: [CPU:dec, Heap: dec, RAM: dec, NW: dec, Disk: dec]
	C: [CPU:inc, Heap: inc, RAM: inc, NW: inc, Disk: inc]
```

#### First we create an impact matrix for the cluster:
* node A has only decrease across each dim
* B has 1 inc and 1 dec action, 
* C has 1 inc action for each dim

```ruby
	       A 	B 	      C
cpu 	  i=0,d=1     i=1,d=1     i=1,d=0
heap 	  i=0,d=1     i=1,d=1     i=1,d=0
ram       i=0,d=1     i=1,d=1     i=1,d=0
nw        i=0,d=1     i=1,d=1     i=1,d=0
disk      i=0,d=1     i=1,d=1     i=1,d=0
```

#### Now we run the actions through this matrix again.
> If an action is increasing pressure on a dim on a node, it is only allowed 
> if there is no other action trying to dec that dim on the node. i.e. d=0 for that node,dim.

#### When an action gets purged in this second pass, we remove it from the matrix by reducing the node,dim counters.

In above e.g. we process `MoveShard(A->B)` - 
* It cannot go through because `d=1` for some dims on node B
* We purge `MoveShard(A->B)`
* Counters on A and B are updated. All counters in A become `i=0, d=0`. B becomes `i=0,d=1`

Now we process `MoveShard(B->C)` - 
* It increases pressure on C but there is no one decreasing pressure (`d=0` for all dims on C), so it can go through 
* On B it is dec pressure anyway so it has a clear pass.

Hence `MoveShard(B->C)` gets picked.

This will also resolve cycles like A->B, B->C, C->A by picking either one or two of the actions (depending on order of 2nd pass).

---

It is okay if we handle multi-node actions in a separate PR later. You may also use different/better algorithms for polarizing impact vectors. Just wanted to check if the current `Collator::polarize()` and `ActionGrouper::groupByInstanceId()` is the right approach for future use cases, like the moveShard, splitShard etc..

Reply:
Thanks for this sage advice, I had an unnecessarily complex system and I was able to rip it out until I got to something similar to what you've proposed here :) I've updated the PR to reflect these changes.